Grid map unification

mishrasubhransu opened the pull request:
Used shared_ptr instead of pointers.
Used vectors<> instead of arrays
Merged grid and map into GridCore class and added more function
CarGeometry has a safety buffer around it
Added better visualization for car and primitives.
Templated CarConnectivity, simplified it as well.
CarCost handles all obstacles avoidance and gives consistent cost for true cost and heur cost.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jhu-asco/dsl/5)
<!-- Reviewable:end -->

mishrasubhransu committed:
Removed bug in flow and added CarConnectivity with twist(2d) primitives

mishrasubhransu committed:
modified grid.cc/h and cargrid.cc/h to handle wrapped dimension

mishrasubhransu committed:
replaced all pointers with shared_ptr

mishrasubhransu committed:
few more stl related changes and addition of se2 operations to utils

mishrasubhransu committed:
common template class for grid and map, better plots for paths in bin, work on all example, used header only cereal serialization

mishrasubhransu committed:
Changed comments and replaced CarGrid slice with GridCore slice

mishrasubhransu committed:
cleanup

mishrasubhransu committed:
added option for reversing connectivity when graph expanded backwards and cleanup

mishrasubhransu committed:
primitive plotting functionality in connectivity, modified cost in carconnectivity to take turning into accout

mishrasubhransu committed:
Paralellzed cmap creation, added CarCost to CarConnectivity so that the cost is consistent with heuristics

mishrasubhransu committed:
added methods to GridCore to stackup grids, methods to save all the slices of grids(cmap) to files and added files that were missing in last commit

mishrasubhransu committed:
obstacle checking and cost of primitive handled by CarCost

mishrasubhransu committed:
added LoopOver function(and speed tests) to GridCore thats faster than calling grid.Index() on every cell

mishrasubhransu committed:
cleaned up the Id, Index, Get, Set functions of GridCore Class

mishrasubhransu committed:
removed unnecessary asserts from grid.h, merged and tested CarConnectivity classes

mishrasubhransu committed:
merged CarPrimitiveConfig into connectivity class and merged two cartests into one

mishrasubhransu committed:
changed the lanes.cfg etc to use twistnorm in carcost

mishrasubhransu committed:
rolled out loops in Index() and LoopOver() for efficiency and removed storage of grid index for each cell

mishrasubhransu committed:
added template method(using variadic templates) to find if template argument is shared_ptr to enable or disable data copy

mishrasubhransu committed:
get returned zero which was only ok for pointers and arithmetic types, changed it to return default object

mishrasubhransu committed:
terrain planner support

marinkobi commented on CMakeLists.txt:
```diff
@@ -39,6 +42,9 @@ add_subdirectory (bin)
         DESTINATION lib/pkgconfig)
 
+#Install the header only serialization library called cereal
+INSTALL(DIRECTORY third_party/cereal DESTINATION include)
```
I though we were going to use protobuf?

marinkobi commented on bin/cartest.cc:
```diff
@@ -21,32 +51,48 @@ vector<Vector2d> ToVector2dPath(const CarPath &path) {
   Params params(argv[1]);
 
+  //To plot the path as rectanges or just points
+  bool plot_car; params.GetBool("plot_car", plot_car);
```
split in two lines

marinkobi commented on bin/cartest.cc:
```diff
@@ -55,11 +101,15 @@ int main(int argc, char** argv)
-  Vector4d whxy;
-  bool useGeom = params.GetVector4d("geom", whxy);
-  
+  Vector5d whxy;
```
5d?  rename variable to reflect this update

marinkobi commented on bin/cartest.cc:
```diff
@@ -55,11 +101,15 @@ int main(int argc, char** argv)
   // load an occupancy map from ppm file
-  dsl::Map<bool, 2> omap = load(mapName.c_str(), ocs.tail<2>());
+  dsl::Map<bool, 2>::Ptr pomap = load(mapName, ocs.tail<2>());
+  dsl::Map<bool, 2>& omap = *pomap;
```
please use only one variable here, i.e. just omap as a shared pointer, othewise it's confusing

marinkobi commented on bin/cartest.cc:
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
     if (useGeom) {
-      CarGeom geom(whxy(0), whxy(1), whxy(2), whxy(3));
-      CarGrid::MakeMap(geom, omap, *cmap);
+      int nthreads; params.GetInt("nthreads", nthreads);
```
splint into two lines

marinkobi commented on bin/cartest.cc:
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
-  timer_start(&timer);
+  double ac;
+  Vector3d wt;
+  shared_ptr<CarCost> cost;
```
this doesn't need to be a shared pointer. You can just set it as an object passed by reference.

marinkobi commented on bin/cartest.cc:
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
+      cout << "Map, with primitives from start saved to prim2.ppm" << endl;
+    }
+
+  }
```
these are very good additions! 
one thing though is that this file used to be much shorter and simpler, and now it's gotten too long. We should factorize it so that all the path plotting and the new stuff happens as part of the library not as part of this test. 
The tests ideally should be short and should not introduce a lot of new functionality

marinkobi commented on lib/carconnectivity.h:
```diff
@@ -52,13 +77,79 @@ class CarConnectivity : public GridConnectivity< Eigen::Vector3d, Eigen::Matrix3
+   * @param cfg the configuration for generation of primitives
+   * @return
+   */
+  bool SetPrimitives(double dt, CarPrimitiveConfig& cfg){
```
this should be in a *.cc file

mishrasubhransu replied:
This is a template class, so shouldn't the implementation stay in the .h file?

marinkobi replied:
OK, I hand't noticed that it's now a template class -- we should use templates for sparingly, i.e. only for the most generic implementation -- carconnectivity is too specific and definitely should not be a templated class -- can we remove the dependence on ConnectionType and implement it differently?

mishrasubhransu replied:
I have different connection types and different data type, for which most of the code is identical.  What would be the best way to implement is without code duplication?

marinkobi commented on lib/carconnectivity.h:
```diff
@@ -80,36 +171,91 @@ class CarConnectivity : public GridConnectivity< Eigen::Vector3d, Eigen::Matrix3
-  bool Flow(std::tuple< SE2Cell*, SE2Path, double>& pathTuple,
-            const Eigen::Matrix3d& g0,
-            const Eigen::Vector3d& v) const;
+  bool GetPrims(const Vector3d pos, vector<vector<Vector2d>>& prims ){
```
should not be in header

mishrasubhransu replied:
This is a template class, so shouldn't the implementation stay in the .h file?

marinkobi commented on lib/carcost.cc:
```diff
@@ -2,23 +2,80 @@
-  return (1 - eps)*(d1 + ac * d2);
-}
+  if(use_twistnorm_metric){
+    return (1 - eps)*(twist.array()* wt.array()).matrix().norm();
```
that looks awkward, don't you just need an dot product?

mishrasubhransu replied:
I need the a norm(scaled by the weight wt) of the twist, so dot product doesn't work.

marinkobi commented on lib/carcost.cc:
```diff
@@ -2,23 +2,80 @@
-  double d1 = (a.c.tail< 2 >() - b.c.tail< 2 >()).norm(); // position distance
-  double d2 = fabs(fangle(a.c(0) - b.c(0))); // orientation distance
-  return d1 + ac * d2;
+double CarCost::Real(const SE2Cell& a, const SE2Cell& b) const{
```
this function looks expensive, where is it used?  
the real cost ideally should be computed incrementally at the same time when one integrates the edges (e.g. inside Connectivity)

mishrasubhransu replied:
It is indeed done incrementally and the Real() is only called from inside Connectivity. It's basically what the connectivity used to do. It made sense to have both the costs together to make sure that the heuristics is admissible.

marinkobi commented on lib/cargeom.cc:
```diff
@@ -0,0 +1,76 @@
+      points.push_back(Eigen::Vector2d(x + ox_, y + oy_));
+}
+
+double CarGeom::l() const {return l_;}
```
it's ok to define these inline in the header

marinkobi commented on lib/cargrid.cc:
```diff
@@ -11,198 +11,30 @@
+  //Iterate over all cells
+  auto fun = [&](int id, const Vector3i& gidx){
+    Vector3d cc; //CellCenter
+    bool gotcenter = CellCenter(cc,gidx);assert(gotcenter);
```
split lines

marinkobi commented on lib/cargrid.h:
```diff
@@ -13,34 +13,13 @@
-#include "map.h"
 
 namespace dsl {
+using std::shared_ptr;
```
not a good idea to include, just use std:: inline

marinkobi commented on lib/cell.h:
```diff
@@ -16,8 +16,10 @@ namespace dsl {
  * a point in the cell.
  */
-template < class PointType, class DataType >
+template < class PT, class DT >
```
that's confusing, please revert to old syntax

mishrasubhransu replied:
I do it because I typedef them again as PointType and DataType so that it is accessible for use elsewhere like so
`using CarTwistConnectivity     = SE2GridConnectivity<SE2Cell::PointType, SE2Cell::DataType, SE2Twist>;`
`using CarPathConnectivity      = SE2GridConnectivity<SE2Cell::PointType, SE2Cell::DataType, SE2Path>;`

What do you suggest I do?

marinkobi replied:
how about SE2Cell::Point ?

marinkobi commented on lib/cargrid.h:
```diff
@@ -13,34 +13,13 @@
-};
-
+using SE2Cell =  Cell< Eigen::Vector3d, Eigen::Matrix3d >; // a cell that stores an SE(2) transformation  matrix
+using SE2CellPtr = shared_ptr<SE2Cell>;
```
using shared_ptr for cells is a bad idea... they are very expensive and should only be used for large objects that truly require shared access and have no clearly defined notion of who owns them

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
 namespace dsl {
 
 struct EmptyData {};
+using std::shared_ptr;
```
we should not be including "using std::"

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
  */
-template < class PointType, class DataType = EmptyData>
- struct Grid {
+template < class PointType, class CellContent>
```
For consistency one should use Type for the second argument, why change the naming here?

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+   * @return the cell array id. -1 if point x is not in grid
+   */
+  int Id(const Vectornd& x) const {
+    Vectorni gidx; Index(gidx,x);
```
split lines

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+   */
+  int Id(const Vectorni& gidx) const {
+    if(Valid(gidx))
+      return gidx.transpose()*cgs;
```
the goal  was to make this more efficient by explicitly avoiding the transpose and multiplication for low-dimensional cases, why is it removed?

mishrasubhransu replied:
I benchmarked against the old code(with checks for out of bound) and found the following elapsed times for 1e7 calls:

grid2.gs():10 10
dt Id:0.113313 secs
dt IdOld:0.0857138 secs

grid3.gs():10 10 10
dt Id:0.170228 secs
dt IdOld:0.127534 secs

grid4.gs():10 10 10 10
dt Id:0.220866 secs
dt IdOld:0.16996 secs

grid5.gs():10 10 10 10 10
dt Id:0.230863 secs
dt IdOld:0.212158 secs

Also gidx.transpose()*cgs is faster than gidx.dot(cgs)

marinkobi replied:
yes, I would say 30% gain in efficiency is significant

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+   */
+  int Index(double xi, int i) const {
+    if(wd[i]){ //dimension is wrapped
+      while(xi < xlb[i]){xi += ds[i];}
```
using while here looks dangerous, is the loop guaranteed to execute at most twice? If not one should use mod()

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+    for(size_t i=0;i<n;i++){
+      double xi = x[i];
+      if(wd[i]){ //dimension is wrapped
+        while(xi < xlb[i]){xi += ds[i];}
```
same as above

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+        break;
+      default:
+        Vectornp1i gidxe = Vectornp1i::Zero();//e stands for extra element
+        Vectornp1i  gse; gse << gs,0; //0 to indicate iteration over all cells over
```
split lines

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+   */
+  void ToGridCoordinates(vector<Vectornd>& coord) const{
+
+    Vectornd point0; CellCenter(point0,Vectorni::Zero());
```
split

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
+ * e.g. dim=5 or 6.
+ */
+template < class PointType, class DataType = EmptyData>
+using Grid = GridCore< PointType,shared_ptr< Cell<PointType, DataType> > >;
```
not a good idea to use shared_ptr here

marinkobi commented on lib/gridsearch.h:
```diff
@@ -60,9 +60,11 @@ using std::vector;
-template < class PointType, class DataType, class ConnectionType = std::vector< Cell<PointType, DataType>* > >
-  class GridSearch : public Search< Cell<PointType, DataType>, ConnectionType > {
+template < class PT, class DT, class ConnectionType = std::vector< Cell<PT, DT>* > >
+  class GridSearch : public Search< Cell<PT, DT>, ConnectionType > {
```
please revert to the old names, these are not descriptive at all

marinkobi commented on lib/gridutils.cc:
```diff
@@ -0,0 +1,398 @@
+  return false;
+}
+
+//bool MakeSE2Map(const CarGeom& geom, const Map<bool, 2> &omap, Map<bool, 3> &cmap, int nthreads) {
```
remove if not needed

marinkobi commented on lib/grid.h:
```diff
@@ -10,178 +10,941 @@
-template < class PointType, class DataType = EmptyData>
- struct Grid {
+template < class PointType, class CellContent>
+class GridCore {
```
Why change the name to GridCore? Isn't this just a Grid? 
using Grid = GridCore< PointType,shared_ptr< Cell<PointType, DataType> > >; 
is confusing since Grid according to it becomes an alias for GridCore (that's contrary to thinking of Core as something residing inside the Grid).  
I would just call GridCore -> Grid
and then use a more specific alias, e.g.
SE2Grid  = Grid< Eigen::Vector3d, SE2Cell::Ptr>
or something like that -- i.e. all names should be consistent with the template arguments

Again, super important -- ::Ptr here should never be a shared pointer, consider using unique pointers instead

mishrasubhransu replied:
There were two different classes Grid and Map before, instantiated differently e.g. Grid<Vector3d, Matrix3d> and Map<3,bool> . While the cells in Grid stored Cell<Vector3d, Matrix3d>* , the cells in Map stored bools. 

To retain the same structure, I create this GridCore and aliased Grid and Map from it. Please advice how should  I structure them both finally.

marinkobi commented:
That's a lot of good functionality! 
We need to be a bit more careful in making sure that these new choices do not effect efficiency (see specific comments above). Also, the executable tests should be as simple as possible. They also serve as an illustration how to setup and run a problem, and if they're too long and complex it's hard to follow the logic. So it's good to have e.g. a super-simple cartest.cc and a more complex e.g. carviztest.cc with fancy visualization, but better yet factor out the plotting/visualization and call it with a line or two of code in all tests. 
Also, any chance we can use protobuf instead of cereal?

marinkobi commented:
Also, we haven't been too strict about this, but any new functionality should follow
https://google.github.io/styleguide/cppguide.html
, since it looks like you have decided to follow some of the google-style approach, but not all of it, so we should be more consistent and just enforce it from now on. 

Also, I'm not sure that using vectors instead of arrays is a good idea, have you done benchmarking on it? The goal of using an array is because we know its constant size in advance and so it can be pre-allocated. Vectors ops can be more expensive in general.

mishrasubhransu committed:
terrain planner working, added methods to read .ppm files reliably and easily, bugs

mishrasubhransu committed:
missed example for reading writing ppm

mishrasubhransu committed:
made data members of  private or protected, added accessors and mutators and modified few things according to google styleguide.

mishrasubhransu committed:
changed Functions to CamelCase

mishrasubhransu committed:
more naming change, moved ppm related stuff to separate file

mishrasubhransu committed:
changed header guard as per standard

mishrasubhransu committed:
removed cereal and used protobuf

mishrasubhransu committed:
works with protobuf, lots of comments in examples ensured all abs() are std::abs()

mishrasubhransu committed:
disabled save and load for grid containing pointers

mishrasubhransu commented:
I made a new branch(google_styleguide) with the changes that you suggested above. Uses protobuf, follows google styleguide(new stuff) and more. I simplified the test examples and added good explanation at the beginning. 

I used vectors because it made life a lot easier. I haven't benchmarked it but, everyone at [stackoverflow](http://stackoverflow.com/questions/381621/using-arrays-or-stdvectors-in-c-whats-the-performance-gap) seems to be strongly suggesting to use vectors instead of using new. My vectors are allocated once in the constructors and after that the size can't be changed. If you would like I can do some benchmarking to see the performance difference. 

Lastly, you said in the above comment "We need to be a bit more careful in making sure that these new choices do not effect efficiency (see specific comments above)". I was unable to find any specific comments.

mishrasubhransu committed:
extended grid saving beyond vector<bools> to any type and also shared_ptr, cleaned up examples

mishrasubhransu committed:
styleguide changes

mishrasubhransu committed:
made non pod methods to have stream serialization and parsing compatible with protobuf. Consumes lot more memory though

mishrasubhransu committed:
removed bugs in serialization and deserialization

mishrasubhransu committed:
modified function arguments according to google style guide and marins reviews

mishrasubhransu committed:
minor corrections

mishrasubhransu committed:
some typedefs to make it read easier

mishrasubhransu committed:
last commit didn't compile

mishrasubhransu committed:
more typedefs

mishrasubhransu committed:
replaced while loop for angle wrapping

mishrasubhransu committed:
bug in typedef

mishrasubhransu committed:
using pointer instead of smart_ptr

mishrasubhransu committed:
back to using share_ptr

mishrasubhransu committed:
typedefs

mishrasubhransu committed:
Renamed GridCore to Grid

mishrasubhransu committed:
introduced a delete method for grid

mishrasubhransu committed:
memory allocation for grid(holding ptr) takes place inside grid

mishrasubhransu committed:
incomplete

mishrasubhransu committed:
incomplete

mishrasubhransu committed:
works using unique_ptr and returning null reference

mishrasubhransu committed:
changes to method names in grid.h

mishrasubhransu committed:
only using unique_ptr to allocate cells. cell allocation optional

mishrasubhransu committed:
not using null reference any more, returning pointer

mishrasubhransu committed:
faster Id method

mishrasubhransu committed:
updated docs

mishrasubhransu committed:
removed unused variable from Grid

mishrasubhransu committed:
some minor refactoring

mishrasubhransu committed:
some visualization updates

mishrasubhransu committed:
made cargrid cell center at 0, 90, 180 and 270

mishrasubhransu committed:
plotting forward and reverse car with different color

mishrasubhransu committed:
some cleanup and visualization changes

mishrasubhransu committed:
removed some commented code

mishrasubhransu committed:
added a new constructor for grid

mishrasubhransu committed:
fixed bug in constructor added in last commit

mishrasubhransu committed:
all warning are errors and fixed a bug in terrainse2grid.cc

mishrasubhransu committed:
Typedefs made consistent in gridsearch.h

mishrasubhransu committed:
fixed bug in terrain planner codes

mishrasubhransu committed:
added automatic compilation of .proto files and removed some redundant variables in carcost

mishrasubhransu committed:
include header changed to quotes

mishrasubhransu committed:
changed terrain map loading and saving. Added barycentre finding class
